### ISSUE TITLE
include `fn centered_rect` from `examples/popup.rs` in API

### ISSUE THREAD

**Issue by wookietreiber** [ratatui/ratatui#617]
**Note:** This is more or less a copy of my issue on the original `tui-rs` project: https://github.com/fdehau/tui-rs/issues/498.

## Problem

I find the `fn centered_rect` in `examples/popup.rs` to be extremely useful, so much, that I would like it to be part of the API. I'd probably copy this function to every TUI project :sweat_smile: 

## Solution

- add the function to the public API, possibly on `Rect` itself, as it always takes a `Rect` as an argument
- keep the `popup.rs` example and use the function from the public API instead of defining it there again

## Alternatives

- keep it as is
- have a `ratatui-util` crate to contain functions like these

## Additional context

Whether to include this in the public API depends on your inclination towards having these kinds of utility functions in the API. I have heard different opinions on this and I'd respect not adding this. IMHO if enough members of the community use this function anyway, I'd include it in the API, as it would be a good way to do cross-project code deduplication and to have a reference to the usecase in the API documentation.

**Comment by joshka** [comment]
@kdheepak wrote in #618

---
Thanks for the PR! I also tend to have this in almost every application, and I believe this is a good candidate for inclusion. 

There is already a `inner` method:

https://github.com/ratatui-org/ratatui/blob/59b9c32fbc2bc6725bdec42e63216024fab71493/src/layout/rect.rs#L93

that people can use like this:

```rust
Rect::new(1, 2, 3, 4).inner(&Margin::new(1, 2)),
```

I'm not sure if this can be done in a non-breaking manner, but it would have been nice to use the same function for different purposes.

e.g.


```rust
Rect::new(1, 1, 10, 10).inner(Margin::px(1, 2))
Rect::new(1, 1, 10, 10).inner(Margin::percent(10, 20))
```

**Comment by joshka** [comment]
>I'm not sure if this can be done in a non-breaking manner, but it would have been nice to use the same function for different purposes.

Regarding non-breaking often replacing a parameter that takes `T` (e.g. Margin) with one that takes `Into<U>` (e.g.  `Into<InnerRectConstraint>`) is a pretty easy way to maintain mostly backwards compatible things.

> There is already a inner method:

I definitely agree with @kdheepak's implication here - centered seems too specific / narrow. For context the definition in #617 is:

```rust
pub fn centered_rect(self, percent_x: u16, percent_y: u16) -> Rect
```

I'm pushing back against using this as the API as it only really solves a narrow use case and I'm not certain that's the one that all users would want. I don't know what an API for this would look like - so let's design it!

It's possible that something using `layout::Constraint` would work
It's possible that overloading `Rect::inner()` or fixing `Margin` would work
It's possible that we might want to just hang this off the layout module directly rather than off Rect (probably not though).

Some things to think about when designing this:
- We currently have two types used for similar things in Ratatui:
  - [layout::Margin](https://docs.rs/ratatui/latest/ratatui/layout/struct.Margin.html) has just horizontal / vertical amount
  - [block::Padding](https://docs.rs/ratatui/latest/ratatui/widgets/block/struct.Padding.html) has left/right/top/bottom
- What are some requirements for layout that we'd want to support?
  - A Rect with a known size centered in another Rect
  - A Rect with an calculated size based on the percentage of the screen (this example)
  - A Rect with an calculated size specific margins above, below, left, right
  - ?

**Comment by wookietreiber** [comment]
I agree that the use-case of the basically copy-pasted `fn centered_rect` from the popup example is too narrow for inclusion in the public API. I'll try whipping something up using generics for us to get a feel for what it'll look like and we can then iterate on that. I'll maybe even do it in a new PR, so we can more easily see the diff in implementation and usage.

**Comment by joshka** [comment]
If you want to just spitball some options for the API, writing some options as code examples of what the user code would look like could work too

**Comment by wookietreiber** [comment]
While experimenting a bit more with the code, I think it might be the better choice to have this functionality live in a separate `layout::Popup` struct to clearly communicate the intent to users of the library.

Even if we find a way to integrate the functionality with generics into `Rect` (or others), I fear it might get too *clever* and thus cryptic. In my opinion, for a library, clearly communicating intent with the API is better than (exaggerating) bending other interfaces to find the one true generic function that does it all, if you know what I mean :sweat_smile: In other words: it's tempting to find clever ways of solving a problem like this but it might not always be the healthy choice for a library.  

Also, when thinking about the `popup.rs` example and the usual use case of clearing the background in advance to drawing the actual popup content/widgets, as in:

```rust
let widget = ...;
let area = Popup::from(frame.size()).percentage(60, 20);
frame.render_widget(Clear, area); // this clears out the background
frame.render_widget(widget, area);
```

... it might even be better turning `Popup` into a slim container widget rather than having it live in the `layout` module. This way, users wouldn't need to have two `render_widget` uses, which would simplify popups on the side of the user. 

---

I'll draft both `layout::Popup` and `widgets::Popup` in new PRs separate from #618.

**Comment by joshka** [comment]
I think a more idiomatic and generally useful implementation for centered might look like the following. I think the size of the `Rect` being a percentage of the total area is an orthogonal concern to the position of the `Rect`.

```rust
impl Rect {
    fn centered(width: u16, height: u16, area: Rect) -> Rect {
        Rect {
            x: area.width.saturating_sub(width) / 2,
            y: area.height.saturating_sub(height) / 2,
            width: min(width, area.width),
            height: min(height, area.height),
        }
    }
}
```

Using this looks like:

```rust
pub struct Popup<'content> {
    pub title: Line<'content>,
    pub body: Text<'content>,
}

fn render(self, area: Rect, buf: &mut Buffer) {
    let height = popup.body.height() as u16 + 2;
    let width = popup.body.width() as u16 + 2;
    let area = centered_rect(width, height, area);
    // -- snip --
}
```

**Comment by Valentin271** [comment]
Do you want to expose this in [tui-popup](https://github.com/joshka/tui-popup) @joshka?

**Comment by joshka** [comment]
I'm not 100% sold on it yet. Perhaps something like:
```rust
Rect::new(0,0,10,10).centered_within(frame.size());
```
is possible another approach that could work.

**Comment by Valentin271** [comment]
Seems a little bit better, maybe we actually need an API to position a `Rect` inside another `Rect`. top-left, bottom-center, center ... all 9 positions ... 

At least that'd be more generic, we could potentially even need that in ratatui (I'm thinking of the chart legend).

**Comment by joshka** [comment]
Interesting, or maybe ...
```rust
rect.position_within(frame.size(), HorizontalAlignment::Center, VerticalAlignment:Top)
```
The `rect.position_within part` seems a little janky. What about?
```rust
rect.inside(frame.size(), HorizontalAlignment::Center, VerticalAlignment:Top)
```

**Comment by joshka** [comment]
I think perhaps it could make sense to have the outside rect create the inside one so you get left to right outer to inner.
```rust
frame.size().create_inner(Rect::new(0,0,10,10, Alignment::Center, VerticalAlignment::Top);
```

**Comment by Valentin271** [comment]
> Interesting, or maybe ...
> 
> ```rust
> rect.position_within(frame.size(), HorizontalAlignment::Center, VerticalAlignment:Top)
> ```

Yep that woks too, that reuses the `Alignment` enum I like that.

It it worth having a type wrapping `Alignment` and `VerticalAlignment`? give a tuple? or stay simple and use two parameters as you did. Will we need that kind of positioning elsewhere?

> The `rect.position_within part` seems a little janky. What about?
> 
> ```rust
> rect.inside(frame.size(), HorizontalAlignment::Center, VerticalAlignment:Top)
> ```

I like that too.

> I think perhaps it could make sense to have the outside rect create the inside one so you get left to right outer to inner.
> 
> ```rust
> frame.size().create_inner(Rect::new(0,0,10,10, Alignment::Center, VerticalAlignment::Top);
> ```

I have no strong opinion to do it that way or the other way around, I like both, but 2 things:
- `Rect::new` is not needed since `x` and `y` will always be dropped, a tuple or two parameters would do, or even a Size struct
- Either that or, still give a rect but then change the `create` part, it's confusing to me because we're not creating anything, we're positioning/placing (or synonym) a rect

Or maybe we'll support positioning a `Rect` with `x` and `y` set as sort of an offset, e.g.
```rust
frame.size().create_inner(Rect::new(2,0,10,10, Alignment::Center, VerticalAlignment::Center);
```
would create a centered rect with an offset of 2 on the right. But then we couldn't offset on the left ...
This seems far fetched and a confusing behavior. I'd opt for simply giving the size.

**Comment by joshka** [comment]
Perhaps then `aligned_within` - the problem is it's difficult to work out which is the outer rect with those words.

I'm happy to let this one bubble for a bit and see what we come up with given some time.

**Comment by Valentin271** [comment]
Agree, lets give it some time to mature.

**Comment by joshka** [comment]
With Layout gaining [Flex](https://docs.rs/ratatui/latest/ratatui/layout/struct.Layout.html#method.flex) this looks like:
```rust
let [centered] = Layout::horizontal([Constraint::Percentage(20)]).flex(Flex::Center).areas(area);
```
Where Constraint::Percentage is whatever you need. I think probably removes much of the need for this to be added to the API as anything special.
